core/schema/module.go

```diff
@@ -316,176 +333,841 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
 	}
 
-	self = self.Clone()
-	self.NameField = cfg.Name
-	self.DependencyConfig = cfg.Dependencies
-	self.SDKConfig = cfg.SDK
-	self.SourceDirectory = sourceDir
-	self.SourceDirectorySubpath = sourceDirSubpath
-	self.DependenciesField = deps
+	return mod.Self, nil
+}
 
-	self.Deps = core.NewModDeps(self.Query, self.Dependencies()).
-		Append(self.Query.DefaultDeps.Mods...)
+func (s *moduleSchema) moduleWithName(ctx context.Context, mod *core.Module, args struct {
+	Name string
+}) (*core.Module, error) {
+	mod, err := mod.WithName(ctx, args.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set module name: %w", err)
+	}
 
-	sdk, err := s.sdkForModule(ctx, self.Query, cfg.SDK, sourceDir, sourceDirSubpath)
+	// update dagger.json
+	mod, err = s.updateModuleConfig(ctx, mod)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to update module config: %w", err)
 	}
+	return mod, nil
+}
 
-	self.GeneratedCode, err = sdk.Codegen(ctx, self, sourceDir, sourceDirSubpath)
+func (s *moduleSchema) moduleWithSDK(ctx context.Context, mod *core.Module, args struct {
+	SDK string
+}) (*core.Module, error) {
+	mod, err := mod.WithSDK(ctx, args.SDK)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to set module sdk: %w", err)
 	}
 
-	self.Runtime, err = sdk.Runtime(ctx, self, sourceDir, sourceDirSubpath)
+	// update dagger.json
+	mod, err = s.updateModuleConfig(ctx, mod)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get module runtime: %w", err)
+		return nil, fmt.Errorf("failed to update module config: %w", err)
 	}
 
-	return self, nil
+	return mod, nil
 }
 
-func (s *moduleSchema) moduleInitialize(ctx context.Context, inst dagql.Instance[*core.Module], args struct{}) (*core.Module, error) {
-	return inst.Self.Initialize(ctx, inst, dagql.CurrentID(ctx))
+func (s *moduleSchema) moduleWithDependencies(ctx context.Context, inst dagql.Instance[*core.Module], args struct {
+	Dependencies []core.ModuleSourceID
+}) (*core.Module, error) {
+	deps, err := collectIDInstances(ctx, s.dag, args.Dependencies)
+	if err != nil {
+		return nil, err
+	}
+
+	mod, err := inst.Self.WithDependencies(ctx, s.dag, deps)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set module dependencies config: %w", err)
+	}
+
+	// update dagger.json
+	mod, err = s.updateModuleConfig(ctx, mod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update module config: %w", err)
+	}
+
+	return mod, nil
 }
 
-type asModuleArgs struct {
-	SourceSubpath string `default:""`
+func (s *moduleSchema) moduleGeneratedSourceDirectory(
+	ctx context.Context,
+	mod *core.Module,
+	args struct{},
+) (*core.Directory, error) {
+	// update dagger.json in case there were changes
+	mod, err := s.updateModuleConfig(ctx, mod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update module config: %w", err)
+	}
+
+	// run sdk codegen
+	mod, err = s.updateCodegenAndRuntime(ctx, mod)
```

Comment:
There used to be a reason for this because I previously implemented the loop checking using IDs that were changing in very subtle ways (as opposed to the normalized `ModuleSource`'s ID, which is how it works now), and the fact that codegen didn't run at the time of `withSource` became load bearing to support that.

But I switched away from that approach due to those subtleties, so yeah I actually think I can simplify codegen to just always run on all the `with*` changes (conditionally on name+sdk being set) and make `GeneratedSourceDirectory` a simple field now.

Made that change and the initial tests passed, letting the whole module test suite run now to double check everything but I think this should be good.

Reply:
Ah, okay, ~1 second after I posted my previous comment the test suite run froze and I remembered the other subtly around all this...

The typescript SDK needs to figure out the name of the module is generating for in it's implementation. Previously, it used the `ModuleConfig` API that was present, but I removed that since it felt extraneous when all the config settings can just be on `Module` (which in turn is viable now since we distinguish between uninitialized and initialized modules).

So the new implementation just does this: https://github.com/sipsma/dagger/blob/e48830cab8a8b4744b1089b14b6055fe0a4afde6/sdk/typescript/runtime/main.go#L63-L63

The problem is that when codegen runs in `Module.withSource` (which indirectly gets called on that line), that results in codegen running, which results in the typescript SDK calling itself through the API and ends in deadlock (interestingly enough)...

---

I think for now I'll probably just go back and not run the codegen in `withSource`, since ultimately there wasn't any harm in that besides the fact that the internal implementation is slightly more indirect to follow (can add a comment clarifying at least).

The fact that this results in deadlock is not great, it should really be an error, but I'm going to have to create a follow up issue for that general problem since it'll be way out of scope of these changes.

Reply:
Addendum, I noticed as part of this that I could update the SDK-as-module implementations to accept `ModuleSource` now (they currently accept `modSource *Directory, subPath string`, which is essentially just a `ModuleSource` in the new APis). I think I'll do that at least.

That also opens the possibility of adding `ModuleSource.name`, which would let you get the name without necessarily converting to a full blown module... I'll try it out and see if it makes any sense.

Reply:
Did the changes to have SDKs accept `ModuleSource` and adding that `ModuleSource.name` field.

Still left the codegen out of `withSource` even though it's technically not needed any more. The point about it being wasteful to run all that when it's not strictly needed yet seems compelling enough justification. Left a comment about this in withSource though to clarify.